### PR TITLE
[ENH] Show result in Contingency Table widget

### PR DESCRIPTION
**Pull request by jurijsteblaj**
##### Issue

biolab/orange3#2966

##### Description of changes

Extracted the table in Confusion Matrix widget into a separate class and added a copy of Confusion Matrix widget which uses that class. Added this table and clustering scores to Contingency Table widget. Also added Selected Data and Data outputs which correspond to selected cells in the table (same as in Confusion Matrix).

##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: orangecontrib/prototypes/widgets/owcontingency.py [hunk 1]
```diff
@@ -15,32 +23,45 @@ class OWContingencyTable(widget.OWWidget):
     priority = 2010
 
     inputs = [("Data", Table, "set_data", widget.Default)]
-    outputs = [("Contingency Table", Table, )]
+    outputs = [("Contingency Table", Table, ),
+               ("Selected Data", Table),
+               (ANNOTATED_DATA_SIGNAL_NAME, Table)]
```
Comment (JakaKokosar) [inline orangecontrib/prototypes/widgets/owcontingency.py]: Use class definitions for Input and Output signals.

Example:
```    
class Inputs:
    data = Input("Data", Table)
```
 You can read more [here](http://orange-development.readthedocs.io/widget.html#input-output-signal-definitions).

File: orangecontrib/prototypes/widgets/owcontingency.py [hunk 2]
```diff
@@ -15,32 +23,45 @@ class OWContingencyTable(widget.OWWidget):
...
+        gui.rubber(self.controlArea)
+
+        box = gui.vBox(self.controlArea, "Scores")
+        self.scores = gui.widgetLabel(box, "\n\n")
 
         self.apply_button = gui.auto_commit(
             self.controlArea, self, "auto_apply", "&Apply", box=False)
 
+        self.tablemodel = QStandardItemModel(self)
+        view = self.tableview = ContingencyTable(self, self.tablemodel)
```
Comment (JakaKokosar) [inline orangecontrib/prototypes/widgets/owcontingency.py]: Not sure why you initialize QStandardItemModel here. You could simplify this by moving it to ContingencyTable class.

File: orangecontrib/prototypes/widgets/owcontingency.py [hunk 3]
```diff
@@ -55,19 +76,56 @@ def set_data(self, data):
                 self.rows = self.feature_model[0]
                 self.columns = self.feature_model[0]
                 self.openContext(data)
+                self.tableview.set_variables(self.rows, self.columns)
+                self.table = contingency_table(self.data, self.columns, self.rows)
+                self.tableview.update_table(self.table.X, formatstr="{:.0f}")
+        else:
+            self.tablemodel.clear()
```
Comment (JakaKokosar) [inline orangecontrib/prototypes/widgets/owcontingency.py]: As per comment above,  ContingencyTable needs a method for clearing data in QStandardItemModel
``` 
self.tableview.clear()
```

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    def __init__(self, parent, tablemodel):
+        super().__init__(editTriggers=QTableView.NoEditTriggers)
+
+        self.classesv = None
+        self.classesh = None
+        self.headerv = None
+        self.headerh = None
+        self.parent = parent
+        self.tablemodel = tablemodel
```
Comment (JakaKokosar) [inline orangecontrib/prototypes/widgets/contingency_table.py]: As discused in comments for OWContingencyTable.

``` 
self.tablemodel = QStandardItemModel() 
```

Comment (JakaKokosar) [reply orangecontrib/prototypes/widgets/contingency_table.py]: And for clearing data
```
def clear(self):
    self.tablemodel.clear()
```

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+            for side, p1, p2 in (("t", rect.topLeft(), rect.topRight()),
+                                 ("r", rect.topRight(), rect.bottomRight()),
+                                 ("b", rect.bottomLeft(), rect.bottomRight()),
+                                 ("l", rect.topLeft(), rect.bottomLeft())):
+                if side in borders:
+                    painter.drawLine(p1, p2)
+            painter.restore()
+
+
+class CircleItemDelegate(BorderedItemDelegate, gui.VerticalItemDelegate):
```
Comment (astaric) [inline orangecontrib/prototypes/widgets/contingency_table.py]: Pleas add a docstrings for this class and all its methods

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    def sizeHint(self, option, index):
+        if index.column() == 0 or index.row() == 1:
```
Comment (astaric) [inline orangecontrib/prototypes/widgets/contingency_table.py]: Could you restructure this to first handle header rows/columns and use QStyleItemDelegates implementation for everything else? As far as I can tell, col 0 and row 1 are vertical, column 1 and row 0 are bordered, everything else is Styled.

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+            return gui.VerticalItemDelegate.sizeHint(self, option, index)
+        elif index.column() == 1:
+            return BorderedItemDelegate.sizeHint(self, option, index)
+        elif 2 <= index.row() and 2 <= index.column():
+            return QStyledItemDelegate.sizeHint(self, option, index)
+        else:
+            return BorderedItemDelegate.sizeHint(self, option, index)
+
+    def paint(self, painter, option, index):
+        area = index.data(CircleArea)
```
Comment (astaric) [inline orangecontrib/prototypes/widgets/contingency_table.py]: Same here, can you restructure the ifs to keep BorderedItemDelegate calls together? A comment on why different implementations are used would also not hurt.

File: orangecontrib/prototypes/widgets/contingency_table.py (hunk 4 above)
Comment (astaric) [reply orangecontrib/prototypes/widgets/contingency_table.py]: I agree with @JakaKokosar on this. ContingencyTable should be in charge of creating the model as it already fills it with custom items.

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+            Horizontal top header.
+        """
+        self.classesv = classesv
+        self.classesh = classesh
+        self.headerv = headerv
+        self.headerh = headerh
+        self.initialize(**kwargs)
+
+    def initialize(self, circles=False):
+        """
```
Comment (astaric) [inline orangecontrib/prototypes/widgets/contingency_table.py]: I would make this method private, set_variables and set_headers should be used by outside callers.

Comment (astaric) [reply orangecontrib/prototypes/widgets/contingency_table.py]: For circles parameter, i would rather add setShowCircles method that changes the value and then call initialize.

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        item.setTextAlignment(Qt.AlignCenter)
+        item.setFlags(Qt.NoItemFlags)
+
+        self._set_item(0, 2, item)
+        item = self._item(2, 0)
+        item.setData(self.headerv, Qt.DisplayRole)
+        item.setTextAlignment(Qt.AlignHCenter | Qt.AlignBottom)
+        item.setFlags(Qt.NoItemFlags)
+        self.setItemDelegateForColumn(0, gui.VerticalItemDelegate())
+        self._set_item(2, 0, item)
```
Comment (astaric) [inline orangecontrib/prototypes/widgets/contingency_table.py]: could you split this method into two or more? (one to create cell items, one for headers) As it is, it is hard to follow what is going on.

File: orangecontrib/prototypes/widgets/contingency_table.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+        selection = QItemSelection()
+        index = self.model().index
+        for row, col in indexes:
+            sel = index(row + 2, col + 2)
+            selection.select(sel, sel)
+        self.selectionModel().select(
+            selection, QItemSelectionModel.ClearAndSelect)
+
+    def update_table(self, matrix, colsum=None, rowsum=None, colors=None, formatstr="{}", tooltip=None):
+        """
```
Comment (astaric) [inline orangecontrib/prototypes/widgets/contingency_table.py]: This method could also be split into several shorter (private) methods. (i'd go with at least one method for creating cells, one for sums).

File: orangecontrib/prototypes/widgets/owcontingency.py (hunk 3 above)
Comment (astaric) [reply orangecontrib/prototypes/widgets/owcontingency.py]: +1

File: orangecontrib/prototypes/widgets/contingency_table.py (hunk 8 above)
Comment (astaric) [reply orangecontrib/prototypes/widgets/contingency_table.py]: Does the contingencyTable allow display of circles? If not, should it?

**Comment by astaric** [comment]
Thanks

**Comment by ajdapretnar** [comment]
It is very bad to have two widgets with the same name. Now there are two Confusion Matrices, one in core and one in Prototypes. Please rename. This would cause a lot of confusion among users, especially during our workshops where we use Prototypes add-on.
